protocol/x/clob/keeper/process_operations.go

```diff
@@ -248,6 +214,218 @@ func (k Keeper) PersistMatchToState(
 	return nil
 }
 
+// PersistOrderRemovalToState takes in an OrderRemoval, statefully validates it according to
+// RemovalReason, and writes the removal to state.
+func (k Keeper) PersistOrderRemovalToState(
+	ctx sdk.Context,
+	orderRemoval *types.OrderRemoval,
+) error {
+	orderIdToRemove := orderRemoval.GetOrderId()
+	orderIdToRemove.MustBeStatefulOrder()
+
+	// Order removals are always for stateful orders that must exist.
+	orderToRemove, err := k.FetchOrderFromOrderId(ctx, orderIdToRemove, make(map[types.OrderId]types.Order, 0))
+	if err != nil {
+		return err
+	}
+
+	// Statefully validate that the removal reason is valid.
+	switch removalReason := orderRemoval.RemovalReason; removalReason {
+	case types.OrderRemoval_REMOVAL_REASON_UNDERCOLLATERALIZED:
+		// For the collateralization check, use the remaining amount of the order that is resting on the book.
+		remainingAmount, hasRemainingAmount := k.MemClob.GetOrderRemainingAmount(ctx, orderToRemove)
+		if !hasRemainingAmount {
+			panic(fmt.Sprintf("addOrderToOrderbookCollateralizationCheck: order has no remaining amount %v", orderToRemove))
```

Comment:
let's return an error here? according to the internal audit checklist, we should only panic when something is really really wrong.